lom/src/engine/engine.go

```diff
@@ -0,0 +1,207 @@
+            /* No more writes. Hence close. Does not block reader. */
+            /* This will help abort read loop below */
+            close(p.chClientReq)
+        }()
+
+        /*
+         * In forever read loop until aborted.
+         * Send read requests via chan to main loop.
+         * Watch abort chan to quit.
+         */
+        for {
+            /* Blocking read until request or error or signal in chClReadAbort */
+            req, err := p.tx.ReadClientRequest(p.chClReadAbort)
+
+            if req != nil {
+                /* Select can block write into p.chClientReq, so watch chClReadAbort too */
+                select {
+                case p.chClientReq <- req:
+                    break
+                case <- p.chClReadAbort:
+                    return
+                }
+            } else {
+                LogError("Failed to read request. err(%v)", err)
+                return
+            }
+        }
+    }()
+}
+
+
+func (p *engine_t) runLoop() {
+    /*
+     * Wait for
+     *      signal to refresh
+     *      request from client
+     *      Internal timer for outstanding request's/seq's timeout processing
+     */
+
+    /* To abort periodic log at the end of loop */
+    chAbortLog := make(chan interface{}, 1)
+    LogPeriodicInit(chAbortLog)
+
+    defer func() {
+        /* Indicate loop end */
+        p.chTrack <- 1
+
+        /* Abort LogPeriodic */
+        chAbortLog <- 0
+
+        if len(p.chClReadAbort) < cap(p.chClReadAbort) {
+            p.chClReadAbort <- "Aborted"
+        }
+
+        close(p.chTrack)
+        close(chAbortLog)
+        close(p.chClReadAbort)
+    }()
+
+    /* Handle signal for config update & terminate */
+    chSignal := make(chan os.Signal, 3)
+    signal.Notify(chSignal, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)
+
+    /*
+     * kick off reading client requests
+     *
+     * p.chClientReq - The read loop sends read requests via this chan to main loop.
+     * p.chClReadAbort - Way to abort the read loop.
+     */
+    p.chClientReq = make(chan *LoMRequestInt, 1)
+    p.chClReadAbort = make(chan interface{}, 2)
+    p.readRequest()
+
+    /*
+     * Initialize seq handler
+     *
+     * The seq handler is invoked for any response from clients via this 
+     * runLoop Go routine. To ensure that any timer based processing too
+     * happens in the same context, it uses this routine for any of its
+     * timer call back needs.
+     * On any async timer firing, it intimates this loop via chSeqHandler channel
+     * and the main loop invokes sequence handler.
+     *
+     * This way any logical processing by sequence handler is via the context
+     * of single Go routine only.
+     */
+    chSeqHandler := make(chan int64, 2)
+    InitSeqHandler(chSeqHandler)
+
+    server := GetServerReqHandler()
+
+    /* Intimate the start of loop */
+    p.chTrack <- 0
+loop:
+    for {
+        select {
+        case msg := <- p.chClientReq:
```

Comment:
what is p.chClientReq is closed ? server.processRequest(msg) is called with invalid msg(default channel value). Also this is in for{}, it can create busy loop.

Reply:
It is closed only by engine_t::readRequest()
WHen closed, reader's continue until no more in the channel. After the close, <- operation blocks. When under select, <- never triggers again.

Reply:
Isn't when channel is closed  "case msg := <- p.chClientReq: " will always be executed. msg will have nil value and server.processRequest(msg) is always called ? 
Receive on a closed channel will return the zero value for the channel's type without blocking

Reply:
This one

Reply:
Yes. Fixed the code.

Doc on close:  [link](https://go.dev/ref/spec#Close) 

No value in closing, unless you use that as indicator to the receiver, via no more value or if the receiver uses range.

Fix: Close engine instead and not closing any channels